### PR TITLE
Message codec for Firebase

### PR DESCRIPTION
**Pull request by Skylled**
Potential solution to flutter/flutter#13043 and related issues.
The idea and initial proof of concept for this is @pauldemarco's doing. I don't take any credit.

This is just an extended version of the StandardMessageCodec for the cloud_firestore plugin.

I've tested it successfully with `DocumentReference`, `timestamp`/`DateTime`, and `GeoPoint`.
~~My test with `CollectionReference` did not go as expected.~~ CollectionReference is not a valid type to embed in a Document.
I'm opening this early for feedback/help.

Todo:
- [x] iOS implementation
- [x] Double-check all documentation
- [x] Tests
- [x] Resolve issue with CollectionReference

### CODE REVIEW

**Comment by Skylled** [comment]
In my test, I have a DocumentReference being set on button press, with a CollectionReference being written into it.
Logs:
```E/MethodChannel#plugins.flutter.io/cloud_firestore(15021): Failed to handle method call
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021): java.lang.RuntimeException: Found conflicting setters with name: setWallpaper (conflicts with setWallpaper defined on android.content.ContextWrapper)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejx.<init>(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zzg(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zzd(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejx.zze(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zzd(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejx.zze(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zzd(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejx.zze(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.android.gms.internal.zzejw.zzbp(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.firebase.firestore.zzk.zzbw(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.google.firebase.firestore.DocumentReference.set(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at io.flutter.plugins.firebase.cloud_firestore.CloudFirestorePlugin.onMethodCall(CloudFirestorePlugin.java:204)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:191)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at io.flutter.view.FlutterNativeView.handlePlatformMessage(FlutterNativeView.java:122)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at android.os.MessageQueue.next(MessageQueue.java:323)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at android.os.Looper.loop(Looper.java:136)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at android.app.ActivityThread.main(ActivityThread.java:6247)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:872)
E/MethodChannel#plugins.flutter.io/cloud_firestore(15021):      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:762)
E/flutter (15021): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (15021): PlatformException(error, Found conflicting setters with name: setWallpaper (conflicts with setWallpaper defined on android.content.ContextWrapper), null)
E/flutter (15021): #0      FirestoreMethodCodec.decodeEnvelope (package:cloud_firestore/src/firestore_codecs.dart:407:7)
E/flutter (15021): #1      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:154:18)
E/flutter (15021): <asynchronous suspension>
E/flutter (15021): #2      DocumentReference.setData (package:cloud_firestore/src/document_reference.dart:26:30)
E/flutter (15021): #3      TestPage._addTestDocs (/data/user/0/io.flutter.plugins.firebase.firestoreexample/cache/exampleGFZLBX/example/lib/main.dart:91:10)
E/flutter (15021): <asynchronous suspension>
E/flutter (15021): #4      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:323:14)
E/flutter (15021): #5      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:375:30)
E/flutter (15021): #6      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:102:24)
E/flutter (15021): #7      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:161:9)
E/flutter (15021): #8      TapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:123:7)
E/flutter (15021): #9      GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:156:27)
E/flutter (15021): #10     BindingBase&SchedulerBinding&GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:147:20)
E/flutter (15021): #11     BindingBase&SchedulerBinding&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:121:22)
E/flutter (15021): #12     BindingBase&SchedulerBinding&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:101:7)
E/flutter (15021): #13     BindingBase&SchedulerBinding&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:64:7)
E/flutter (15021): #14     BindingBase&SchedulerBinding&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:48:7)
E/flutter (15021): #15     _invoke1 (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:115)
E/flutter (15021): #16     _dispatchPointerDataPacket (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:73)```

**Comment by Skylled** [comment]
The Obj-C side of this is outside of my experience, especially without a way to test locally, so I'll continue to work on and perfect the Java side.

**Comment by mit-mit** [comment]
cc @collinjackson

**Comment by Skylled** [comment]
Okay, so I just hit a snag testing an array containing a document reference (something from my personal use-case) which requires we roll Firebase to 11.6.0. [See notes here.](https://groups.google.com/forum/#!topic/google-cloud-firestore-discuss/bk3oX1RIleU)

I'll change all relevant plugins to ensure we don't run into problems like #13324.

**Comment by Skylled** [comment]
I don't know if the tests work or pass, as Travis doesn't seem to be running them anymore. Windows doesn't currently have support for `flutter test`

**Comment by Skylled** [comment]
Thanks to amazing new `flutter test` support on Windows, I was able to run tests and make adjustments as required. This is now 100% on the Dart and Android side.

**Comment by Skylled** [comment]
@collinjackson I just noticed the `ServerValue` in [collection_reference.dart](https://github.com/flutter/plugins/blob/b41fc638b839d3de7c2d04f133cd4b5c27f5fbea/packages/cloud_firestore/lib/src/collection_reference.dart#L44) but I can't find any info about it in Firestore, only Firebase Database. Firestore seems to use [FieldValue](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/FieldValue) instead. [Further evidence](https://groups.google.com/forum/#!topic/firebase-talk/8z8O2jY4fTM) seems to say this class is not correct.

**Comment by Skylled** [comment]
Relevant ObjC code for anyone willing to assist:
[FlutterChannels.h](https://github.com/flutter/engine/blob/2383bb7c4e579d6b13e5b3a988a94d6109b1403a/shell/platform/darwin/ios/framework/Headers/FlutterChannels.h)
[FlutterChannels.mm](https://github.com/flutter/engine/blob/2383bb7c4e579d6b13e5b3a988a94d6109b1403a/shell/platform/darwin/ios/framework/Source/FlutterChannels.mm)
[FlutterCodecs.h](https://github.com/flutter/engine/blob/2383bb7c4e579d6b13e5b3a988a94d6109b1403a/shell/platform/darwin/ios/framework/Headers/FlutterCodecs.h)
[FlutterStandardCodec_Internal.h](https://github.com/flutter/engine/blob/2383bb7c4e579d6b13e5b3a988a94d6109b1403a/shell/platform/darwin/ios/framework/Source/FlutterStandardCodec_Internal.h)
[FlutterStandardCodec.mm](https://github.com/flutter/engine/blob/2383bb7c4e579d6b13e5b3a988a94d6109b1403a/shell/platform/darwin/ios/framework/Source/FlutterStandardCodec.mm)

**Comment by Skylled** [comment]
This is now absolutely as far as I can take this PR. I can't test the iOS code without a Mac.

Could somebody give it a look over? @collinjackson @mit-mit

File: packages/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/FirestoreMessageCodec.java [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+      for (final Entry entry : map.entrySet()) {
+        writeValue(stream, entry.getKey());
+        writeValue(stream, entry.getValue());
+      }
+    } else if (value instanceof Date) {
+      stream.write(DATE_TIME);
+      final Date date = (Date) value;
+      final long milliseconds = date.getTime();
+      writeLong(stream, milliseconds * 1000);
+    } else if (value instanceof GeoPoint) {
```
Comment (Skylled) [inline packages/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/FirestoreMessageCodec.java]: There's no FieldValue encoding here as I'm not sure if there's a way to check whether it's a `serverTimestamp()` or `delete()`. This won't be a big blocker, as it should only affect cached data. FieldValues should be converted by the server, right?

File: packages/cloud_firestore/ios/Classes/FirestoreCodecs.mm [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+    NSNumber seconds = [value timeIntervalSince1970] * 1000;
+    SInt64 n = [seconds longValue];
+    [self writeByte:FirestoreFieldDateTime];
+    [_data appendBytes:(UInt8*)&n length:8];
+  } else if ([value isKindOfClass:[FIRFieldValue]]) {
+    if ([value isKindOfClass:[FSTDeleteFieldValue]]) {
+      SInt32 n = 0;
+      [self writeByte:FirestoreFieldFieldValue];
+      [_data appendBytes:(UInt8*)&n length:4];
+    } else if ([value isKindOfClass:[FSTServerTimestampFieldValue]]) {
```
Comment (Skylled) [inline packages/cloud_firestore/ios/Classes/FirestoreCodecs.mm]: `FSTDeleteFieldValue` and `FSTServerTimestampFieldValue` are pulled direct from [here](https://github.com/firebase/firebase-ios-sdk/blob/bde743ed25166a0b320ae157bfb1d68064f531c9/Firestore/Source/API/FIRFieldValue%2BInternal.h)

**Comment by Skylled** [comment]
Tried to implement FieldValue encoding on Java side today.

Testing a `FieldValue.serverTimestamp` through from Dart to Java and back to Dart gives me a very different error (see logs below).

I've traced the error message [here](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/f2f910fd55d05902f1dd7fabec0024256f75034a/google-cloud-firestore/src/main/java/com/google/cloud/firestore/CustomClassMapper.java#L568). I'll do some more digging to see what's up, but I think this might need to be fixed on the Firebase side. If that's the case, I'll have to remove FieldValue support from this PR.

```
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): Failed to handle method call
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): java.lang.RuntimeException: No properties to serialize found on class com.google.firebase.firestore.FieldValue$zzb
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.android.gms.internal.zzetm$zza.<init>(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.android.gms.internal.zzetm.zzg(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.android.gms.internal.zzetm.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.android.gms.internal.zzetm.zza(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.android.gms.internal.zzetm.zzbv(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.firebase.firestore.zzk.zzcc(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.google.firebase.firestore.DocumentReference.set(Unknown Source)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at io.flutter.plugins.firebase.cloud_firestore.CloudFirestorePlugin.onMethodCall(CloudFirestorePlugin.java:210)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:191)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at io.flutter.view.FlutterNativeView.handlePlatformMessage(FlutterNativeView.java:128)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at android.os.MessageQueue.next(MessageQueue.java:323)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at android.os.Looper.loop(Looper.java:136)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at android.app.ActivityThread.main(ActivityThread.java:6247)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:872)
E/MethodChannel#plugins.flutter.io/cloud_firestore(29782): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:762)
breaking at exception: PlatformException
I/ViewRootImpl(29782): ViewRoot's Touch Event : ACTION_DOWN
I/ViewRootImpl(29782): ViewRoot's Touch Event : ACTION_UP
E/flutter (29782): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (29782): PlatformException(error, No properties to serialize found on class com.google.firebase.firestore.FieldValue$zzb, null)
E/flutter (29782): #0      FirestoreMethodCodec.decodeEnvelope (package:cloud_firestore/src/firestore_codecs.dart:412)
E/flutter (29782): #1      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:154)
E/flutter (29782): <asynchronous suspension>
E/flutter (29782): #2      DocumentReference.setData (package:cloud_firestore/src/document_reference.dart:32)
E/flutter (29782): #3      _TestAppState.build.<anonymous closure> (file:///c:/Users/Kyle/Source/Flutter/Cybertronix/lib/main.dart:107)
E/flutter (29782): #4      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:323)
E/flutter (29782): #5      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:375)
E/flutter (29782): #6      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:102)
E/flutter (29782): #7      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:161)
E/flutter (29782): #8      TapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:94)
E/flutter (29782): #9      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:315)
E/flutter (29782): #10     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:73)
E/flutter (29782): #11     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:101)
E/flutter (29782): #12     BindingBase&GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:143)
E/flutter (29782): #13     BindingBase&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:121)
E/flutter (29782): #14     BindingBase&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:101)
E/flutter (29782): #15     BindingBase&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:64)
E/flutter (29782): #16     BindingBase&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:48)
E/flutter (29782): #17     _invoke1 (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:128)
E/flutter (29782): #18     _dispatchPointerDataPacket (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:85)
```

**Comment by Skylled** [comment]
Assuming the [Google Cloud Java](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/c281a3e4685e35f1e777ade6191dada83d36fd94/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldValue.java) implementation of Firestore's FieldValue is the same as Android's, (which is a safe assumption given the above error message is the same), then the error is 100% valid and irrelevant to this PR. FieldValues need to be given a property of some kind, as the CustomClassMapper will not take empty `Object()`s.

I would file an issue, but I'm unsure of the correct place, as at first glance the Google Cloud Java repo seems to be completely separate from Android and I can't find the Android repo as easily as I could for iOS.

**Comment by ananfang** [comment]
@Skylled 
I am working on iOS (Objective-C) codes. 
When I finish (maybe couple hours later), I'll give you a pull request on your branch `patch-3`

File: packages/cloud_firestore/ios/Classes/FirestoreCodecs.mm (hunk 2 above)
Comment (ananfang) [reply packages/cloud_firestore/ios/Classes/FirestoreCodecs.mm]: @Skylled 
I've just refined iOS part, and give you a pull request on your `patch-3` branch 😸

**Comment by Skylled** [comment]
Due to above comments, I'm going to remove all FieldValue support. We will likely need a very different solution to handle it.

**Comment by ananfang** [comment]
@Skylled 
It's weird. I am sure that I've signed the CLA, but Google bot says `NO`...

**Comment by Skylled** [comment]
@ananfang It's because there's now two authors for this PR, and, as a bot, it isn't sure that you consent to your code being used here. Whoever merges the PR will handle the problem.

**Comment by ananfang** [comment]
@Skylled 
About FieldValue issue, after tracking the latest `cloud_firestore` package (version `0.2.0`). I found there are already `setData with merge option` and `updateData` API.

You started this pull request based on 0.1.x package, if we update to 0.2.0 code base, FiledValue should be supported.

**Comment by Skylled** [comment]
Right, but the Android-side problem can't be fixed from this PR. If a FieldValue is added to any Document, from Flutter code, the app will crash because the cache will attempt to show the raw FieldValue before sending to the server.

I believe this fix relies on a new version of GMS/Firebase to roll out. If a Googler could comment saying this issue is on the radar for the next version, we could stall this PR and put all plugins onto something like 11.8.2 instead of rolling to 11.6.0 now then rolling again shortly after.

Inconsistent GMS/Firebase versions cause other fatal errors in projects.

**Comment by ananfang** [comment]
@Skylled 
For your reference, I tried merging the latest 0.2.0 package and testing FieldValue.delete with setData + SetOptions.merge and updateData, it works without crash.

If you are interested, check branch [update-to-0.2.0-package](https://github.com/ananfang/plugins/tree/update-to-0.2.0-package) on my fork :)

**Comment by Hixie** [comment]
@Skylled What's the status of this PR?

**Comment by Skylled** [comment]
@Hixie Ready to be merged. Or I could also rebase to reduce conflicts with the recent Gradle roll.

The main pain point though, is that all first-party plugins need to roll to GMS 11.6+ which means all plugins will need a new version pushed to Pub.

If you're okay with that, in the rebase or in this PR I could update the changelogs and pubspecs of all affected plugins. [See above comment](https://github.com/flutter/plugins/pull/282#issuecomment-350100213)

**Comment by ananfang** [comment]
@Hixie 
Because @Skylled started this PR before `cloud_firestore 0.2.0`, there are two issues using new Message Codec with `0.1.x`:

1. Lack of **setData with Merge option** and **updateData** API, it will cause FieldValue.delete crash the app.
2. Android dependencies issue (should use Firebase/GMS 11.6.+)

I've tried merge `cloud_firebase 0.2.0` into this PR in [this branch](https://github.com/ananfang/plugins/tree/update-to-0.2.0-package). It works fine in my current project without any crash (tested on iOS).

Since 0.2.0 already updated Firebase/GMS dependencies from `11.4.+` to `[11.4.0,12.0[`, it maybe fixed the problem in Android side.

**Comment by Skylled** [comment]
Closing in favor of #343

**Pull request closed without merging**